<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />



<title>Further Resources</title>

<script src="libs/jquery-1.11.0/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.1/css/bootstrap.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.1/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.1/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.1/shim/respond.min.js"></script>
<!-- Favicon -->
<link rel="shortcut icon" type="image/x-icon" href="img/favicon.ico">

<!-- Google analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-8298649-8', 'auto');
  ga('send', 'pageview');

</script>

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="libs/highlight/default.css"
      type="text/css" />
<script src="libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>


<link rel="stylesheet" href="libs/local/main.css" type="text/css" />
<link rel="stylesheet" href="libs/local/nav.css" type="text/css" />
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img { 
  max-width:100%; 
  height: auto; 
}
</style>
<div class="container-fluid main-container">

<header>
  <div class="nav">
    <a class="nav-logo" href="index.html">
      <!-- <img src="static/img/stat545-logo-s.png" width="70px" height="70px"/> -->
      <img src="img/logo-100px.jpg" width="80px" height="80px"/>
    </a>
    <ul>
      <li class="home"><a href="index.html">Home</a></li>
      <li class="faq"><a href="faq.html">FAQ</a></li>
      <li class="syllabus"><a href="syllabus.html">Syllabus</a></li>
      <li class="topics"><a href="topics.html">Topics</a></li>
      <li class="people"><a href="people.html">People</a></li>
    </ul>
  </div>
</header>

<div id="header">
<h1 class="title">Further Resources</h1>
</div>

<div id="TOC">
<ul>
<li><a href="#r-resources">R resources</a><ul>
<li><a href="#getting-help">Getting Help</a></li>
<li><a href="#general-r-resources">General R Resources</a></li>
<li><a href="#dplyr-resources">dplyr resources</a></li>
<li><a href="#ggplot2-resources">ggplot2 resources</a></li>
<li><a href="#markdown-rmarkdown-resources">Markdown / RMarkdown resources</a></li>
</ul></li>
<li><a href="#rna-seq-resources">RNA-seq resources</a></li>
</ul>
</div>

<div id="r-resources" class="section level1">
<h1>R resources</h1>
<div id="getting-help" class="section level2">
<h2>Getting Help</h2>
<ul>
<li><a href="http://lmgtfy.com/?q=R+Error+in+library+there+is+no+package+called">Google it!</a>: Try Googling generalized versions of any error messages you get. That is, remove text that is specific to your problem (names of variables, paths, datasets, etc.). You’d be surprised how many other people have probably had the same problem and solved it.</li>
<li><a href="http://stackoverflow.com/questions/tagged/r">Stack Overflow</a>: There are over 100,000 questions tagged with “R” on SO. <a href="https://stackoverflow.com/questions/tagged/r?sort=votes">Here are the most popular ones, ranked by vote</a>. Always search before asking, and make a <a href="http://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example/5963610#5963610">reproducible example</a> if you want to get useful advice. This is a minimal example that allows others who are trying to help you to see the error themselves.</li>
<li><a href="https://support.bioconductor.org/">Bioconductor Support Site</a>: Like SO, but specifically for Bioconductor-related questions.</li>
<li>Read package vignettes. For example, see the <a href="https://cran.r-project.org/web/packages/dplyr/index.html">dplyr CRAN page</a>, scroll about halfway down to see the <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/introduction.html">introduction to dplyr</a> vignette.</li>
</ul>
</div>
<div id="general-r-resources" class="section level2">
<h2>General R Resources</h2>
<ul>
<li><a href="http://tryr.codeschool.com/">TryR</a>: An interactive, browser-based R tutor</li>
<li><a href="http://swirlstats.com/">Swirl</a>: An R package that teaches you R (<em>and statistics!</em>) from within R</li>
<li><a href="https://stat545-ubc.github.io/">Jenny Bryan’s Stat 545 “Data wrangling, exploration, and analysis with R” course material</a>: An excellent resource for learning R, dplyr, and ggplot2</li>
<li><a href="https://www.datacamp.com/courses/free-introduction-to-r">DataCamp’s free introduction to R</a></li>
<li><a href="https://www.rstudio.com/resources/cheatsheets/">RStudio’s printable cheat sheets</a></li>
<li><a href="http://rseek.org/">Rseek</a>: A custom Google search for R-related sites</li>
<li>Bioconductor <a href="https://www.bioconductor.org/help/package-vignettes/">vignettes</a>, <a href="https://www.bioconductor.org/help/workflows/">workflows</a>, and <a href="https://www.bioconductor.org/help/course-materials/">course/conference materials</a></li>
</ul>
</div>
<div id="dplyr-resources" class="section level2">
<h2>dplyr resources</h2>
<ul>
<li><a href="https://cran.rstudio.com/web/packages/dplyr/vignettes/introduction.html">The dplyr vignette</a></li>
<li><a href="http://www.dataschool.io/dplyr-tutorial-for-faster-data-manipulation-in-r/">A longer dplyr tutorial with video and code</a></li>
<li><a href="http://genomicsclass.github.io/book/pages/dplyr_tutorial.html">The dplyr tutorial from the HarvardX Biomedical Data Science MOOC</a></li>
<li><a href="https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf">A dplyr cheat sheet from RStudio</a></li>
</ul>
</div>
<div id="ggplot2-resources" class="section level2">
<h2>ggplot2 resources</h2>
<ul>
<li><a href="http://docs.ggplot2.org/">The official ggplot2 documentation</a></li>
<li><a href="http://amzn.to/1akjqsR">The ggplot2 book</a>, edition 1, by the developer, Hadley Wickham</li>
<li><a href="https://github.com/hadley/ggplot2-book">New version of the ggplot2 book, freely available on GitHub</a></li>
<li><a href="https://groups.google.com/d/forum/ggplot2">The ggplot2 Google Group</a> (mailing list, support forum)</li>
<li><a href="http://learnr.wordpress.com/">LearnR</a>: A blog with a good number of posts describing how to reproduce various kind of plots using ggplot2</li>
<li><a href="http://stackoverflow.com/questions/tagged/ggplot2">SO questions tagged with ggplot2</a></li>
<li><a href="http://shinyapps.stat.ubc.ca/r-graph-catalog/">A catalog of graphs made with ggplot2, complete with accompanying R code</a></li>
<li><a href="https://www.rstudio.com/wp-content/uploads/2015/05/ggplot2-cheatsheet.pdf">RStudio’s ggplot2 cheat sheet</a></li>
</ul>
</div>
<div id="markdown-rmarkdown-resources" class="section level2">
<h2>Markdown / RMarkdown resources</h2>
<ul>
<li><a href="http://bioconnector.org/markdown">Basic Markdown + RMarkdown reference</a></li>
<li><a href="https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf">RStudio’s RMarkdown Cheat Sheet</a> and <a href="https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf">RMarkdown Reference Sheet</a></li>
<li><a href="http://rmarkdown.rstudio.com/">RStudio’s RMarkdown documentation</a>: covers all the v2 enhancements to RMarkdown, <a href="http://rmarkdown.rstudio.com/authoring_basics.html">RMarkdown basics</a> in more detail than covered here, the full <a href="http://rmarkdown.rstudio.com/authoring_pandoc_markdown.html">Pandoc Markdown specification</a> which lets you do more than just regular Markdown, information on <a href="http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html">citations and bibliographies</a>, and several <a href="http://rmarkdown.rstudio.com/articles.html">examples</a> illustrating advanced usage.</li>
<li><a href="http://yihui.name/knitr/">knitr documentation</a>: has lots of useful reference material about how knitr works, <a href="http://yihui.name/knitr/options/">options</a>, controlling output, <a href="http://yihui.name/knitr/demos/">examples</a>, and more.</li>
</ul>
</div>
</div>
<div id="rna-seq-resources" class="section level1">
<h1>RNA-seq resources</h1>
<ul>
<li><a href="http://rnaseq.uoregon.edu/">University of Oregon’s RNA-seqlopedia</a>: a comprehensive guide to RNA-seq starting with experimental design, going through library prep, sequencing, and data analysis.</li>
<li><a href="http://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0881-8">Conesa et al. A survey of best practices for RNA-seq data analysis. <em>Genome Biology</em> 17:13 (2016)</a>. If there’s one review to read on RNA-seq and data analysis, it’s this one.</li>
<li><a href="http://rnaseq.wiki">rnaseq.wiki</a> &amp; accompanying <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004393">paper</a> for hands-on RNA-seq data analysis examples using cloud computing.</li>
<li><a href="http://www.rna-seqblog.com/">RNA-seq blog</a>: Several blog posts per week on new methods and tools for RNA-seq analysis.</li>
<li><a href="https://www.youtube.com/playlist?list=PLfFNmoa-yUIb5cYG2R1zf5rtrQQKZvKwG">YouTube playlist: 2015 UC Davis Workshop on RNA-seq methods &amp; algorithms (Harold Pimentel)</a>.</li>
<li><a href="https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/">What the FPKM? A review of RNA-Seq expression units</a> A blog post from Harold Pimentel describing the relationship between R/FPKM and TPM.</li>
<li><a href="https://usegalaxy.org/u/jeremy/p/galaxy-rna-seq-analysis-exercise">RNA-seq analysis exercise using Galaxy</a>: an example analysis you can run yourself using the Tophat+Cufflinks workflow.</li>
<li><a href="http://f1000research.com/articles/4-1070/v1">“RNA-Seq workflow: gene-level exploratory analysis and differential expression.”</a> This paper walks through an end-to-end gene-level RNA-Seq differential expression workflow using Bioconductor packages, starting from FASTQ files.</li>
<li><a href="http://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">The DESeq2 paper</a> describes the modeling approach and shows some benchmarks against other normalization and differential expression strategies.</li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf">The DESeq2 vignette</a> is packed full of examples on using DESeq2, importing data, fitting models, creating visualizations, references, etc.</li>
<li><a href="http://arxiv.org/abs/1104.3889">Lior Pachter’s paper “Models for transcript quantification from RNA-Seq”</a> reviews different approaches for quantifying expression from RNA-seq data and how these affect downstream analysis.</li>
<li>SEQAnswers <a href="http://seqanswers.com/forums/forumdisplay.php?f=26">RNA-seq</a> and more general <a href="http://seqanswers.com/forums/forumdisplay.php?f=18">bioinformatics</a> forums are a great place to search for answers.</li>
<li><a href="https://www.biostars.org/t/RNA-Seq/">Biostars RNA-seq</a> Q&amp;A section.</li>
<li><a href="http://www.gettinggeneticsdone.com/2015/12/tutorial-rna-seq-differential.html">Blog post</a> and <a href="https://dx.doi.org/10.6084/m9.figshare.1619655.v1">printable PDF</a> I created demonstrating how to do pathway analysis with RNA-seq data and R.</li>
</ul>
</div>

<div class="footer">
This work is licensed under the  <a href="http://creativecommons.org/licenses/by-nc/3.0/">CC BY-NC 3.0 Creative Commons License</a>.<br>
For more information, visit <a href="https://www.bioconnector.virginia.edu/">bioconnector.virginia.edu</a>.<br>
<a href="https://twitter.com/genetics_blog" target="_blank"><i class="fa fa-twitter fa-lg"></i></a>&nbsp;
<a href="https://github.com/stephenturner/" target="_blank"><i class="fa fa-github fa-lg"></i></a>&nbsp;
<a href="http://www.gettinggeneticsdone.com/" target="_blank"><i class="fa fa-rss fa-lg"></i></a>&nbsp;
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
