<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />



<title>Reproducible Reporting Homework</title>

<script src="libs/jquery-1.11.0/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.1/css/bootstrap.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.1/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.1/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.1/shim/respond.min.js"></script>
<!-- Favicon -->
<link rel="shortcut icon" type="image/x-icon" href="img/favicon.ico">

<!-- Google analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-8298649-8', 'auto');
  ga('send', 'pageview');

</script>

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="libs/highlight/default.css"
      type="text/css" />
<script src="libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>


<link rel="stylesheet" href="libs/local/main.css" type="text/css" />
<link rel="stylesheet" href="libs/local/nav.css" type="text/css" />
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img { 
  max-width:100%; 
  height: auto; 
}
</style>
<div class="container-fluid main-container">

<header>
  <div class="nav">
    <a class="nav-logo" href="index.html">
      <!-- <img src="static/img/stat545-logo-s.png" width="70px" height="70px"/> -->
      <img src="img/logo-100px.jpg" width="80px" height="80px"/>
    </a>
    <ul>
      <li class="home"><a href="index.html">Home</a></li>
      <li class="faq"><a href="faq.html">FAQ</a></li>
      <li class="syllabus"><a href="syllabus.html">Syllabus</a></li>
      <li class="topics"><a href="topics.html">Topics</a></li>
      <li class="people"><a href="people.html">People</a></li>
    </ul>
  </div>
</header>

<div id="header">
<h1 class="title">Reproducible Reporting Homework</h1>
</div>

<div id="TOC">
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#data-analysis">Data analysis</a><ul>
<li><a href="#load-and-examine-biological-processes">Load and examine biological processes</a></li>
<li><a href="#visualize-genes-in-a-particular-biological-process">Visualize genes in a particular biological process</a></li>
</ul></li>
</ul>
</div>

<p><strong><em>INSTRUCTIONS</em></strong>: Following the horizontal lines below, beginning with the <strong>Introduction</strong> heading, is a report that is produced from an RMarkdown document. Write an RMarkdown document and “Knit HTML” to produce a report similar to what you see below.</p>
<ol style="list-style-type: decimal">
<li>Start by clicking on the menus: <em>File</em>, <em>New File</em>, <em>R Markdown…</em>.</li>
<li>Choose <em>HTML document</em>, and give your document a title and an author’s name.</li>
<li>Using the appropriate Markdown syntax and embedding RMarkdown chunks, reproduce the report below. (Feel free to copy and paste text as needed).
<ul>
<li>Ensure that the hyperlink to the paper in the Introduction section works properly.</li>
<li>Ensure that the <strong>bold</strong>, <em>italicized</em>, and <code>fixed width</code> text come through in your document. Fixed width text is produced by enclosing text in backtick characters (the key above the tab key, to the left of the number 1 key on most keyboards).</li>
</ul></li>
<li>Ensure that this report is fully reproducible. You can assume that any computer this runs on will have the appropriate libraries installed (dplyr, readr, ggplot2). Notice that in the <em>Load and examine biological processes</em> subheading under the <em>Data analysis</em> heading uses <code>read_csv()</code> to read data <em>directly from the web</em>, rather than from a local file. Thus, for this code to work, you must have internet access.</li>
</ol>
<hr />
<p><mark><strong><em>THE REPORT THAT YOU SHOULD PRODUCE FOLLOWS BELOW.</em></strong></mark></p>
<hr />
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This data is a tidy version of a gene expression dataset from <a href="http://www.ncbi.nlm.nih.gov/pubmed/17959824">Brauer et al. Coordination of Growth Rate, Cell Cycle, Stress Response, and Metabolic Activity in Yeast (2008) <em>Mol Biol Cell</em> 19:352-367</a>. This data is from a gene expression microarray, and in this paper the authors are examining the relationship between growth rate and gene expression in yeast cultures limited by one of six different nutrients (glucose, leucine, ammonium, sulfate, phosphate, uracil). If you give yeast a rich media loaded with nutrients except restrict the supply of a <em>single</em> nutrient, you can control the growth rate to any rate you choose. By starving yeast of specific nutrients you can find genes that:</p>
<ol style="list-style-type: decimal">
<li><strong>Raise or lower their expression in response to growth rate</strong>. Growth-rate dependent expression patterns can tell us a lot about cell cycle control, and how the cell responds to stress. The authors found that expression of &gt;25% of all yeast genes is linearly correlated with growth rate, independent of the limiting nutrient. They also found that the subset of negatively growth-correlated genes is enriched for peroxisomal functions, and positively correlated genes mainly encode ribosomal functions.</li>
<li><strong>Respond differently when different nutrients are being limited</strong>. If you see particular genes that respond very differently when a nutrient is sharply restricted, these genes might be involved in the transport or metabolism of that specific nutrient.</li>
</ol>
</div>
<div id="data-analysis" class="section level2">
<h2>Data analysis</h2>
<p>Below we will load the dataset and proceed with a very basic analysis.</p>
<div id="load-and-examine-biological-processes" class="section level3">
<h3>Load and examine biological processes</h3>
<p>After loading the <strong>dplyr</strong> and <strong>readr</strong> libraries, use readr’s <code>read_csv()</code> function to read the data directly from the web, assigning it to an object in our workspace called <code>ydat</code>.</p>
<pre class="r"><code>library(dplyr)
library(readr)
ydat &lt;- read_csv(&quot;http://bioconnector.org/data/brauer2007_tidy.csv&quot;)</code></pre>
<p>Which biological processes have the most genes? That is,</p>
<ol style="list-style-type: decimal">
<li>Take the ydat data,</li>
<li><em>then</em> group it by biological process,</li>
<li><em>then</em> summarize it to count the number of distinct genes within each process,</li>
<li><em>then</em> arrange it descending by the largest number of genes,</li>
<li><em>then</em> show only the top 20 results.</li>
</ol>
<pre class="r"><code>ydat %&gt;% 
  group_by(bp) %&gt;% 
  summarize(n=n_distinct(symbol)) %&gt;% 
  arrange(desc(n)) %&gt;% 
  head(20)</code></pre>
<pre><code>## Source: local data frame [20 x 2]
## 
##                                                              bp     n
##                                                           (chr) (int)
## 1                                    biological process unknown   269
## 2                                          protein biosynthesis   182
## 3                           protein amino acid phosphorylation*    78
## 4                                         protein biosynthesis*    73
## 5                        cell wall organization and biogenesis*    64
## 6  regulation of transcription from RNA polymerase II promoter*    49
## 7                        nuclear mRNA splicing, via spliceosome    47
## 8                                                   DNA repair*    44
## 9                                          aerobic respiration*    42
## 10                                       ER to Golgi transport*    42
## 11                                                 endocytosis*    41
## 12                                         mRNA-nucleus export*    41
## 13                                          response to stress*    39
## 14                                  protein-vacuolar targeting*    37
## 15                        cell wall organization and biogenesis    36
## 16                                             rRNA processing*    35
## 17                       ubiquitin-dependent protein catabolism    35
## 18                                        ER to Golgi transport    33
## 19                           35S primary transcript processing*    30
## 20                             chromatin silencing at telomere*    29</code></pre>
</div>
<div id="visualize-genes-in-a-particular-biological-process" class="section level3">
<h3>Visualize genes in a particular biological process</h3>
<p>Create a new object that contains the data filtered down to the 36 genes involved in <em>cell wall organization and biogenesis</em>. Plot a scatter plot of the expression versus rate, color-coding by the limiting nutrient, showing a different gene in each panel, where the y-axis scale in each panel can vary from panel to panel.</p>
<pre class="r"><code># First load the ggplot2 library
library(ggplot2)

cwob &lt;- ydat %&gt;% filter(bp==&quot;cell wall organization and biogenesis&quot;)

ggplot(cwob, aes(rate, expression, color = nutrient)) +
  geom_point() +
  geom_smooth(method = &quot;lm&quot;, se = FALSE) +
  facet_wrap(~symbol, scales=&quot;free_y&quot;) + 
  theme_bw() + 
  xlab(&quot;Growth rate&quot;) + 
  ylab(&quot;Gene expression (log-normalized ratio)&quot;) + 
  ggtitle(&quot;Expression of genes involved in cell wall organization and biogenesis&quot;)</code></pre>
<p><img src="r-hw-repres_files/figure-html/cellwall_org_plots-1.png" alt="" /></p>
<hr />
<p><mark><strong><em>THE REPORT YOU SHOULD PRODUCE ENDS HERE</em></strong></mark></p>
<hr />
</div>
</div>

<div class="footer">
This work is licensed under the  <a href="http://creativecommons.org/licenses/by-nc/3.0/">CC BY-NC 3.0 Creative Commons License</a>.<br>
For more information, visit <a href="https://www.bioconnector.virginia.edu/">bioconnector.virginia.edu</a>.<br>
<a href="https://twitter.com/genetics_blog" target="_blank"><i class="fa fa-twitter fa-lg"></i></a>&nbsp;
<a href="https://github.com/stephenturner/" target="_blank"><i class="fa fa-github fa-lg"></i></a>&nbsp;
<a href="http://www.gettinggeneticsdone.com/" target="_blank"><i class="fa fa-rss fa-lg"></i></a>&nbsp;
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
